src/executionTimingsReporter.ts

```diff
@@ -0,0 +1,139 @@
+}
+
+export interface StepStartedEventsEmitter {
+    on(event: 'stepStarted', listener: (args: StepStartedEventArguments) => void): this;
+    on(event: 'milestoneReached', listener: (args: MilestoneReachedEventArguments) => void): this;
+}
+
+export interface NavigatedToUserRequestedUrlEventsEmitter {
+    on(event: 'navigatedToUserRequestedUrl', listener: () => void): this;
+    once(event: 'navigatedToUserRequestedUrl', listener: () => void): this;
+}
+
+export class StepProgressEventsEmitter extends EventEmitter {
+    constructor(private readonly _nestedEmitters: [StepStartedEventsEmitter] = [] as [StepStartedEventsEmitter]) {
+        super();
+    }
+
+    public emitStepStarted(stepName: string): void {
+        this.emit(stepStartedEventName, { stepName: stepName } as StepStartedEventArguments);
+    }
+
+    public emitMilestoneReached(milestoneName: string): void {
+        this.emit(milestoneReachedEventName, { milestoneName: milestoneName } as MilestoneReachedEventArguments);
+    }
+
+    private subscribeToAllNestedEmitters(event: string, listener: Function): void {
+        for (const nestedEventEmitter of this._nestedEmitters) {
+            nestedEventEmitter.on(event as any, listener as any);
+        }
+    }
+
+    public on(event: string, listener: Function): this {
+        super.on(event, listener);
+        this.subscribeToAllNestedEmitters(event, listener);
+        return this;
+    }
+}
+
+/* Use to track the time executing each step during launch
+    Usage:
+        this.Events.emitStepStarted("Attach");
+        this.Events.emitStepStarted("Attach.AttachToTargetDebuggerWebsocket");
+        this.Events.emitStepStarted("ClientRequest.setBreakpoints");
+        this.Events.emitStepStarted("WaitingAfter.ClientRequest.setBreakpoints");
+        this.Events.emitStepStarted("ClientRequest.setBreakpoints");
+        this.Events.emitStepStarted("WaitingAfter.ClientRequest.setBreakpoints");
+        reporter.generateReport() // Returns the report. Do not call any more methods after this
+
+    The report/telemetry generated looks like this:
+        All	5271.022216 // Time since the reporter is created (which is basically when we launched the debug adapter itself) until we navigated to the user's web page
+        BeforeFirstStep	7.833247 // Time since the reporter is created (which is basically when we launched the debug adapter itself) until we get the first request
+        ClientRequest.initialize	[1.261372] // Time we spent processing the initialize request
+        WaitingAfter.ClientRequest.initialize	[74.939561] // Time we spent after processing the initialize request for the client to send us another request
+        ClientRequest.launch	[2.748096] // Time we spent processing the first part of the launch request until we actually start launching the target debugee .exe
+        LaunchTarget.LaunchExe	10.276666 // The time it takes to spawn the  the target debugee .exe (We launch the .exe but we don't wait for it)
```

Comment:
Why are some arrays and some not?

Reply:
I forgot to update that comment.
"All" and milestones are single values. All the steps are now arrays.

Reply:
Some steps like Attach.RequestDebuggerTargetsInformation or setBreakpoints can happen multiple times. If that happens we send [<timeTakenInFirstOcurrence>, <timeTakenInSecondOcurrence>]. For simplicity we assume that all "steps" can have multiple ocurrences